File: website/src/utils/relativeTime.ts
/**
 * Formats how long ago `from` happened, relative to `now` (defaults to the
 * current time).
 *
 * Sub-minute durations collapse to "< 1 min" so the label never displays a
 * second-level granularity that ticks on every render. Larger durations round
 * down to the largest whole unit: minutes, hours, days, weeks, then years.
 * This mirrors the same helper on iOS (`RelativeTime`) and Android
 * (`RelativeTime`) so the three clients read identically.
 *
 * Returns an empty string when `from` can't be parsed into a date.
 */
export function formatRelativeTime(from: Date | string, now: Date = new Date()): string {
  const fromMs = typeof from === 'string' ? Date.parse(from) : from.getTime()
  if (Number.isNaN(fromMs)) return ''

  const seconds = Math.max(0, Math.floor((now.getTime() - fromMs) / 1000))
  if (seconds < 60) return '< 1 min'

  const minutes = Math.floor(seconds / 60)
  if (minutes < 60) return `${minutes} min`

  const hours = Math.floor(minutes / 60)
  if (hours < 24) return `${hours} hr`

  const days = Math.floor(hours / 24)
  if (days < 7) return `${days} ${days === 1 ? 'day' : 'days'}`

  if (days < 365) {
    const weeks = Math.floor(days / 7)
    return `${weeks} ${weeks === 1 ? 'week' : 'weeks'}`
  }

  const years = Math.floor(days / 365)
  return `${years} ${years === 1 ? 'year' : 'years'}`
}

File: android/PositiveOnlySocial/app/src/main/java/com/example/positiveonlysocial/util/RelativeTime.kt
package com.example.positiveonlysocial.util

import java.util.Date

/**
 * Formats how long ago [from] happened, relative to [now] (defaults to the
 * current time).
 *
 * Sub-minute durations collapse to "< 1 min" so the label never shows a
 * second-level granularity that ticks on every recomposition. Larger durations
 * round down to the largest whole unit: minutes, hours, days, weeks, then
 * years. This mirrors the same helper on iOS (`RelativeTime`) and the website
 * (`formatRelativeTime`) so the three clients read identically.
 */
object RelativeTime {
    fun format(from: Date, now: Date = Date()): String {
        val seconds = ((now.time - from.time) / 1000).coerceAtLeast(0)
        if (seconds < 60) return "< 1 min"

        val minutes = seconds / 60
        if (minutes < 60) return "$minutes min"

        val hours = minutes / 60
        if (hours < 24) return "$hours hr"

        val days = hours / 24
        if (days < 7) return "$days ${if (days == 1L) "day" else "days"}"

        if (days < 365) {
            val weeks = days / 7
            return "$weeks ${if (weeks == 1L) "week" else "weeks"}"
        }

        val years = days / 365
        return "$years ${if (years == 1L) "year" else "years"}"
    }
}
